### ISSUE TITLE
Create json schema for strategeus inputs

### ISSUE THREAD

**Issue by ablack3** [OHDSI/Strategus#9]
My understanding is that Strategus will redefine the json representation for OHDSI studies. 

I think that clearly specifying the JSON format for Strategus is quite important and I would like to suggest that we create a JSON-schema https://json-schema.org/ or something similar that we can use to validate Strategus inputs.

Strategus takes two json files as input: Analysis specifications and Execution Settings

An example Analysis specification is at https://github.com/OHDSI/Strategus/blob/main/extras/cgCdAnalysisSpecifications.json

An example Execution settings json is at https://github.com/OHDSI/Strategus/blob/main/extras/cgCdExecutionSettings.json

**Comment by ablack3** [comment]
Here is a simple example schema for a "concept" (or is it called a "concept expression"?)

``` r
schema <- '
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://ohdsi.org/concept.schema.json",
  "title": "concept",
  "description": "An OMOP concept expression",
  "type": "object",
  "properties": {
    "concept": {
      "description": "An OMOP concept",
      "type": "object",
      "properties": {
        "CONCEPT_CLASS_ID": {
          "type" : "string"
        },
        "CONCEPT_CODE": {
          "type" : "string"
        },
        "CONCEPT_ID": {
          "type" : "integer",
          "minimum": 0
        },
        "CONCEPT_NAME": {
          "type" : "string"
        },
        "DOMAIN_ID": {
          "type" : "string"
        },
        "INVALID_REASON": {
          "type" : "string"
        },
        "INVALID_REASON_CAPTION": {
          "type": "string"
        },
        "STANDARD_CONCEPT": {
          "type": "string",
          "enum": ["S", "N", "C"]
        },
        "STANDARD_CONCEPT_CAPTION": {
          "type": "string"
        },
        "VOCABULARY_ID": {
          "type": "string"
        }
      },
      "required": ["CONCEPT_CLASS_ID", "CONCEPT_CODE", "CONCEPT_ID", "CONCEPT_NAME", "DOMAIN_ID", "INVALID_REASON", "INVALID_REASON_CAPTION", "STANDARD_CONCEPT", "STANDARD_CONCEPT_CAPTION", "VOCABULARY_ID"]
    },
    "isExcluded": {
      "description": "Should this concept be excluded from the concept set?",
      "type": "boolean"
    },
    "includeDescendants": {
      "description": "Should descendents be included/excluded? (true or false)",
      "type": "boolean"
    },
    "includeDescendants": {
      "description": "Should descendents be included/excluded? (true or false)",
      "type": "boolean"
    }
  },
  "required": [ "concept" ]
}
'

validator <- jsonvalidate::json_validator(schema, engine = "ajv")

jsonToValidate <- '
 {
  "concept": {
    "CONCEPT_CLASS_ID": "Clinical Finding",
    "CONCEPT_CODE": "41309000",
    "CONCEPT_ID": 201612,
    "CONCEPT_NAME": "Alcoholic liver damage",
    "DOMAIN_ID": "Condition",
    "INVALID_REASON": "V",
    "INVALID_REASON_CAPTION": "Valid",
    "STANDARD_CONCEPT": "S",
    "STANDARD_CONCEPT_CAPTION": "Standard",
    "VOCABULARY_ID": "SNOMED"
  },
  "includeDescendants": true
}'

validator(jsonToValidate)
#> [1] TRUE
```

<sup>Created on 2022-06-19 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by schuemie** [comment]
Hi @ablack3 ! Yes, we're thinking of how to specify the input to Strategus. I've been playing around with JSON-schema as well, as you can see [here](https://github.com/schuemie/JsonSchemaTest). You can view an example JSON Scheme for the CohortDiagnosticsModule [here](https://json-schema.app/view/%23?url=https%3A%2F%2Fraw.githubusercontent.com%2Fschuemie%2FJsonSchemaTest%2Fmain%2FCohortDiagnosticSpecifications.json).

In R we'd need functions to validate the JSON (`jsonvalidate` seems a good candidate!), and functions to unserialize the JSON into whatever R objects we need. Shouldn't be too difficult. 

An open question is what would generate the JSON. It could become part of ATLAS, in which case it might be nice to have the ability to generate Java code stubs from the JSON Schema (which I think is possible).

**Comment by ablack3** [comment]
We're thinking alike! I learned the the jsonschema package does support references to external json files (using $ref) but does yet support URL references. 

I'd vote for Atlas as json generator. Having solid and stable json schemas will help with collaboration between the java and R side. circe-be would be a good place to start for someone who wants to work on json schemas since I don't think we have circe-be schemas (as far as I know) and those are stable.

**Comment by schuemie** [comment]
Good point. I'm also pretty sure the Circe JSON schema will be the most complicated one.

**Comment by gkovaig** [comment]
Hi @schuemie and @ablack3, Raj Manickam here (https://github.com/gkovaig).  Not sure if this item was discussed and assigned at the Hack-a-thon.

Given that the schema for `strategus` is an aggregation of schema defined in other components (including `CohortGenerator`, `circe-be`, etc.) it makes sense to come up with a set of specifications that cross-reference each other.

From this [note](https://github.com/cran/jsonvalidate/blob/372e767ebd3baf7778f2dc23bb4abeaf5515f67e/R/validate.R#L5) it appears that `jsonvalidator` can use an alternate engine `another json validator` that supports many more features, including nested schemas that span multiple files, meta schema versions later than draft-04, validating using a subschema, and validating a subset of an input data object.

**Comment by schuemie** [comment]
Thanks Raj. Creating the JSON schema for the Strategus modules and collecting them in one place is high on our priority list (I had hoped we could work on this during the hackathon, but we needed to have a discussion on overall Strategus architecture first)

**Comment by ablack3** [comment]
@anthonysena we should take up this topic in the strategus working group. It also relates to some of the work going on with Arachne and specifying execution environments like this one: https://storage.googleapis.com/arachne-datanode/descriptor_base.json

**Comment by anthonysena** [comment]
Noting that something similar has been done for renv: https://github.com/rstudio/renv/pull/1889. This may be applicable to validating the structure of the analysis specification if the renv.lock file is moved into the analysis specification per #144.
